Submit tx endpoint

v0d1ch opened the pull request:
## Why

Our dear colleague @abailly-iohk had the idea to add possibility for the users to submit their transactions to the L1 network using their hydra-node. Our users also confirmed this would, indeed be a nice feature to have thus this PR.


---

<!-- Consider each and tick it off one way or the other -->
* [x] CHANGELOG updated or not needed
* [x] Documentation updated or not needed
* [x] Haddocks updated or not needed
* [x] No new TODOs introduced or explained herafter

v0d1ch committed:
Draft new types needed for the rest endpoint

v0d1ch committed:
Introduce a failing test

v0d1ch committed:
Specialize the submit request to cardano tx

v0d1ch committed:
Implement bare minimum server handler to get a green test

v0d1ch committed:
Add new chain handle to tackle user tx submission

v0d1ch committed:
Call submitTransaction directly in chain handler

v0d1ch committed:
Try to catch submit tx errors and display them to the user

v0d1ch committed:
Test sending signed tx to the endpoint

v0d1ch committed:
Expect SubmitTransactionException in api handler

v0d1ch committed:
Test also that unsigned tx is rejected

v0d1ch committed:
Use chain handle to post user transaction

- The proper way to submit a tx is to use a chain handle. We are able
to catch exceptions in the server code and inform the user + they will
be able to see the possible failures in the logs.
- Remove text output from the server response and return 400 instead of
  500 statuses since this is related to user errors

v0d1ch committed:
Add new types to our api schema

v0d1ch committed:
Match on string contained in the response body

v0d1ch committed:
Improve on a schema wording a bit

v0d1ch committed:
Move APIServerLog to RestServer module

v0d1ch committed:
Move api handlers and server code from Server to RestServer

v0d1ch committed:
Resolve now expired todo

v0d1ch committed:
Changelog entry

ch1bo commented on hydra-node/src/Hydra/API/RestServer.hs:
```diff
@@ -104,13 +145,139 @@ instance Arbitrary DraftCommitTxRequest where
+          }
+      respond $ responseLBS status400 [] "Resource not found"
+
+-- *- Handlers
```
```suggestion
-- * Handlers
```

ch1bo commented on hydra-node/src/Hydra/API/RestServer.hs:
```diff
@@ -104,13 +145,139 @@ instance Arbitrary DraftCommitTxRequest where
+  parseJSON = withText "SubmittedTxResponse" $ \t ->
+    if t == submittedTxResponseConst
+      then pure SubmittedTxResponse
+      else fail $ "Failed to parse: " <> show t <> " expected: " <> show submittedTxResponseConst
```
Should: not serialize a data type into a string, even if it's just a singleton

Besides the fact that we do not need to send ANY body if we don't want to, I don't think that a specific string ("Transaction Submitted") is better than `{"tag": "TransactionSubmitted"}` (I would rename the data constructor)

ch1bo commented on hydra-node/src/Hydra/API/RestServer.hs:
```diff
@@ -104,13 +145,139 @@ instance Arbitrary DraftCommitTxRequest where
+          Left (e :: PostTxError Tx) ->
+            responseLBS status400 [] (Aeson.encode . Aeson.String . pack $ show e)
+          Right _ ->
+            responseLBS status200 [] (Aeson.encode SubmittedTxResponse)
```
Could: Not send anything on success is also fine.

v0d1ch replied:
I want the user to receive something so it is not confusing if the operation succeeded or not, so `TransactionSubmitted` is fine for me.

ch1bo commented on hydra-node/src/Hydra/API/RestServer.hs:
```diff
@@ -104,13 +145,139 @@ instance Arbitrary DraftCommitTxRequest where
+          Right commitTx ->
+            responseLBS status200 [] (Aeson.encode $ DraftCommitTxResponse commitTx)
+ where
+  return400 = responseLBS status400 [] . Aeson.encode . toJSON
```
Could: DRY with line 217 and also use this in the other handlers.

ch1bo commented on hydra-node/src/Hydra/API/RestServer.hs:
```diff
@@ -104,13 +145,139 @@ instance Arbitrary DraftCommitTxRequest where
+  case Aeson.eitherDecode' body :: Either String DraftCommitTxRequest of
+    Left err ->
+      respond $ responseLBS status400 [] (Aeson.encode $ Aeson.String $ pack err)
+    Right requestInput@DraftCommitTxRequest{utxoToCommit} -> do
```
Could: DRY the request parsing + error response on failure

ch1bo commented on hydra-node/src/Hydra/Chain.hs:
```diff
@@ -200,6 +200,8 @@ data Chain tx m = Chain
   -- ^ Create a commit transaction using user provided utxos (zero or many) and
   -- information to spend from a script. Errors are handled at the call site.
+  , submitUserTx :: MonadThrow m => Tx -> m ()
+  -- ^ Submit a signed transaction on behalf of the user.
```
Must: document what exceptions may be thrown by `MonadThrow` (the ones which made us add this constraint)

See https://github.com/input-output-hk/hydra/wiki/Coding-Standards#exceptions

v0d1ch replied:
Hmm...I  added it yesterday but probably my local amend did not work..

ch1bo commented on hydra-node/src/Hydra/Chain/CardanoClient.hs:
```diff
@@ -69,21 +69,39 @@ buildTransaction ::
+  -- | Outputs to create.
+  [TxOut CtxTx] ->
+  IO (Either TxBodyErrorAutoBalance BalancedTxBody)
+buildBalancedTxBody networkId socket changeAddress utxoToSpend collateral outs = do
```
Why are you factoring this out?

I'll comment on the call site how you can simply use the `TxBody` (of the function above)

ch1bo commented on hydra-node/golden/ReasonablySized SubmittedTxResponse.json:
This is a bit pointless.. don't you think? :D

ch1bo commented on hydra-cluster/src/Hydra/Cluster/Scenarios.hs:
```diff
@@ -339,6 +342,61 @@ canCloseWithLongContestationPeriod tracer workDir node@RunningNode{networkId} hy
+      case eTxBody of
+        Left e -> failure $ show e
+        Right body -> do
+          let unsignedTx = unsafeBuildTransaction $ balancedTxBodyContent body
```
Should: just use `makeSignedTransaction` or `signShelleyTransaction` with an empty list on the `TxBody`

ch1bo commented on hydra-node/src/Hydra/API/RestServer.hs:
```diff
@@ -6,29 +6,70 @@ module Hydra.API.RestServer where
+          <$> arbitrary
+          <*> arbitrary
+          <*> oneof [pure Nothing, pure $ Just (Aeson.Object mempty)]
+      ]
```
Why Is the API Server log in the "RestServer" part of the Hydra.API?

I think this is a smell. Move it to a separate module or break the traces down to make things combine better.

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Handlers.hs:
```diff
@@ -172,6 +173,11 @@ mkChain tracer queryTimeHandle wallet@TinyWallet{getUTxO} ctx LocalChainState{ge
               else pure $ Left SpendingNodeUtxoForbidden
           _ -> pure $ Left FailedToDraftTxNotInitializing
+    , -- Post a signed transaction on behalf of the user. Errors are handled at
+      -- the call site and even though they can be tecnically any of
```
```suggestion
      -- the call site and even though they can be technically any of
```

ffakenz commented on hydra-cluster/src/Hydra/Cluster/Scenarios.hs:
```diff
@@ -339,6 +342,61 @@ canCloseWithLongContestationPeriod tracer workDir node@RunningNode{networkId} hy
+    aliceChainConfig <- chainConfigFor Alice workDir nodeSocket [] $ UnsafeContestationPeriod 100
+    let hydraNodeId = 1
+    withHydraNode tracer aliceChainConfig workDir hydraNodeId aliceSk [] [1] hydraScriptsTxId $ \_ -> do
+      -- let's prepare a _user_ transaction from Bob to Carol
```
i think this comment is wrong or not needed

v0d1ch replied:
what I wanted to say is that we are preparing a tx to send something from Bob to Carol and bob here is a hydra-node _user_ using this new feature if that makes sense :) .

ch1bo commented on hydra-node/src/Hydra/API/RestServer.hs:
```diff
@@ -104,13 +145,139 @@ instance Arbitrary DraftCommitTxRequest where
+      handleDraftCommitUtxo directChain tracer body (requestMethod req) (pathInfo req) respond
+    ("GET", ["protocol-parameters"]) ->
+      respond $ responseLBS status200 [] (Aeson.encode pparams)
+    ("POST", ["submit-user-tx"]) -> do
```
Should: find a more REST-ful API

Having verbs in the path is suboptimal. `POST /tx-submission`, `POST /transaction`, `POST /cardano-transaction`, `POST /cardano-tx`, `POST /user-transaction` are all better.

If you want to keep "user", then we should also document what a user transaction is and how it is different to non user transactions.

ffakenz commented on hydra-node/src/Hydra/API/RestServer.hs:
minor: I have a re-naming suggestion, let me know wdyt:
1. I think HttpServer is a better for this module.
2. Then, maybe, we can move the WsServer into its own module as well.
3. Finally the Server can be called TCPServer.

v0d1ch replied:
Yeah, sounds good to me.

ch1bo commented on hydra-cluster/src/Hydra/Cluster/Scenarios.hs:
```diff
@@ -371,3 +429,29 @@ headIsInitializingWith expectedParties v = do
+    L.responseStatus response == toEnum stat && not (B.null chunk) && assertBodyContains mbodyContains chunk
+   where
+    -- NOTE: The documentation says: Response body parameter MAY include the beginning of the response body so this can be partial.
+    -- https://hackage.haskell.org/package/http-client-0.7.13.1/docs/Network-HTTP-Client.html#t:HttpExceptionContent
```
I think this is a bit of a problem of using `req`, if we would be using `http-client` or `http-conduit` (for `Network.HTTP.Simple`), we could use `parseRequest` which does not fail on us and we could inspect the full response body:
https://hackage.haskell.org/package/http-conduit-2.3.8.3/docs/Network-HTTP-Simple.html#v:parseRequest

ch1bo replied:
Maybe a red bin item?

v0d1ch replied:
https://github.com/orgs/input-output-hk/projects/25/views/2?pane=issue&itemId=34471450

ffakenz commented on hydra-node/src/Hydra/Chain/CardanoClient.hs:
```diff
@@ -69,21 +69,39 @@ buildTransaction ::
+  [TxIn] ->
+  -- | Outputs to create.
+  [TxOut CtxTx] ->
+  IO (Either TxBodyErrorAutoBalance BalancedTxBody)
```
nice comments! 👏

ch1bo commented on hydra-node/json-schemas/api.yaml:
```diff
@@ -155,6 +155,34 @@ channels:
+      summary: Submitted user transaction response.
+      operationId: submittedTxResponse
+      message:
+        $ref: "#/components/messages/SubmittedTxResponse"
```
Can't we inline the definition of these? Especially the response is super trivial here (may even be an empty object/null)

ch1bo reviewed (changes requested):
Some refactoring could be avoided and I think we should change the response serialization + API endpoint to something more REST-ful.

v0d1ch committed:
Update hydra-node/src/Hydra/API/RestServer.hs

Co-authored-by: Sebastian Nagel <ch1bo@users.noreply.github.com>

v0d1ch committed:
Update hydra-node/src/Hydra/Chain/Direct/Handlers.hs

Co-authored-by: Sebastian Nagel <ch1bo@users.noreply.github.com>

v0d1ch committed:
Rename SubmitTxResponse to TransactionSubmitted

- And fix the api schema

v0d1ch committed:
Reuse return400 function

v0d1ch committed:
Clarify what errors submitUserTx should return

v0d1ch committed:
Inline messages in the api schema

v0d1ch committed:
Restore CardanoClient

v0d1ch committed:
Rename path to "cardano-transaction"

v0d1ch committed:
Rename RestServer to HTTPServer

- Also rename the corresponding test module
RestServerSpec -> HTTPServerSpec

v0d1ch committed:
Extract the websocket server out of API.Server

This also merges withAPIServer and runAPIServer together as most
complexity comes from arguments being passed in.

v0d1ch committed:
Move api logging type to APIServerLog module

v0d1ch committed:
Revert "Inline messages in the api schema"

This reverts commit 870d22c43ffb59b595844fafbd29cb7be5ca68ae.

v0d1ch committed:
Alter test to use makeSignedTransaction

ffakenz commented on hydra-node/src/Hydra/API/APIServerLog.hs:
```diff
@@ -0,0 +1,44 @@
+  | APIInvalidInput {reason :: String, inputReceived :: Text}
+  | APIConnectionError {reason :: String}
+  | APIHandshakeError {reason :: String}
+  | APIRestInputReceived
```
I think after the renaming this also needs to be refined

ch1bo replied:
Agreed, the data constructors feel a bit off here.

Things which strike me:
* All start with API, which is maybe not needed (the type is about API and usually we have context, only maybe qualified imports are needed to avoid overlaps)
* `RestInput` is rather a `HttpRequest` or just `Request`
  - Why are there multiple `paths`? An http request always has only one path
* The logs with "connection" are really about the websocket connections and could be misinterpreted with general connections (each request comes with a connection)

ffakenz committed:
Fix behavior broken links by using pathname:/// for generated haddock docs

ch1bo commented on hydra-cluster/src/Hydra/Cluster/Scenarios.hs:
```diff
@@ -63,6 +72,7 @@ import Network.HTTP.Req (
  )
 import qualified PlutusLedgerApi.Test.Examples as Plutus
 import Test.Hspec.Expectations (shouldBe, shouldThrow)
+import Test.Hydra.Prelude (failure)
```
Could: put prelude imports in the top block

ch1bo commented on hydra-cluster/src/Hydra/LogFilter.hs:
```diff
@@ -8,7 +8,7 @@ import Hydra.Prelude hiding (id)
 import qualified Data.Map as Map
 import Hydra.API.ClientInput (ClientInput (NewTx))
-import Hydra.API.ServerOutput (ServerOutput (..))
+import Hydra.API.TCPServerOutput (ServerOutput (..))
```
What is the rationale for this renaming?

The original idea was to have a module named by its main data type "ServerOutput". If you really want to do this rename, there are some links to fix to the respective haddocks.

Could: not rename it?

ch1bo commented on hydra-cluster/test/Test/EndToEndSpec.hs:
```diff
@@ -150,6 +151,11 @@ spec = around showLogsOnFailure $
+        withClusterTempDir "submit-a-signed-user-transaction" $ \tmpDir -> do
+          withCardanoNodeDevnet (contramap FromCardanoNode tracer) tmpDir $ \node ->
+            publishHydraScriptsAs node Faucet
+              >>= canSubmitUserTransaction tracer tmpDir node
```
What is a user transaction?

ch1bo commented on hydra-node/exe/hydra-node/Main.hs:
```diff
@@ -4,8 +4,8 @@ module Main where
 
-import Hydra.API.Server (Server (Server, sendOutput), withAPIServer)
-import Hydra.API.ServerOutput (ServerOutput (PeerConnected, PeerDisconnected))
+import Hydra.API.TCPServer (Server (Server, sendOutput), withAPIServer)
```
What's the rationale of calling this a TCP server?

I would expect a barebones TCP connection server which speaks a custom protocol seeing this module name.

But in fact, it's just an HTTP-based API server.

Could: just not rename it?

ch1bo replied:
I have undone the `Server` -> `TCPServer` renaming.

ch1bo commented on hydra-node/golden/ReasonablySized SubmittedTxResponse.json:
Must: not commit unused golden files

This is likely an old one and you accidentally committed it?

ch1bo commented on hydra-node/json-schemas/api.yaml:
```diff
@@ -155,6 +155,34 @@ channels:
           method: GET
           bindingVersion: '0.1.0'
+  /cardano-transaction:
+    description: Submit signed user transaction to be submitted to the L1 network.
```
What is a user transaction?

ch1bo replied:
@v0d1ch I have reworked the HTTP endpoint schema for this and this seems to be the way to go with AsyncAPI currently: https://github.com/input-output-hk/hydra/commit/32430063f2ee65f62a919418d9dc3dc9b16440e9

ch1bo commented on hydra-node/src/Hydra/API/APIServerLog.hs:
```diff
@@ -0,0 +1,44 @@
+import qualified Data.Aeson as Aeson
+import Data.ByteString.Short ()
+import Hydra.Chain.Direct.State ()
+import Hydra.Ledger.Cardano ()
```
Are these imports really needed? I don't see where there would be the need for a Cardano instance?

ch1bo commented on hydra-cluster/src/Hydra/Cluster/Scenarios.hs:
```diff
@@ -339,6 +339,67 @@ canCloseWithLongContestationPeriod tracer workDir node@RunningNode{networkId} hy
+          -- sending SIGNED transaction should be accepted
+          res <- sendRequest hydraNodeId signedRequest
+
+          responseBody res `shouldBe` TransactionSubmitted
```
We could be a bit more succinct here:

```suggestion
          let unsignedRequest = SubmitTxRequest{txToSubmit = makeSignedTransaction [] body}
          sendRequest hydraNodeId unsignedRequest `shouldThrow` expectErrorStatus 400 (Just "MissingVKeyWitnessesUTXOW")

          let signedRequest = SubmitTxRequest{txToSubmit = signTx cardanoBobSk unsignedTx}
          (sendRequest hydraNodeId signedRequest <&> responseBody) `shouldReturn` TransactionSubmitted
```

ch1bo commented on hydra-cluster/src/Hydra/Cluster/Scenarios.hs:
```diff
@@ -339,6 +339,67 @@ canCloseWithLongContestationPeriod tracer workDir node@RunningNode{networkId} hy
+              ReferenceScriptNone
+      -- prepare fully balanced tx body
+      let eTxBody = buildRaw (fst <$> UTxO.pairs bobUTxO) [carolsOutput, bobsOutput] (Lovelace 165545)
+      case eTxBody of
```
Could: case directly or use lambdacase to avoid need for `eTxBody` variable

ch1bo commented on hydra-cluster/src/Hydra/Cluster/Scenarios.hs:
```diff
@@ -339,6 +339,67 @@ canCloseWithLongContestationPeriod tracer workDir node@RunningNode{networkId} hy
+      let carolsAddress = mkVkAddress networkId cardanoCarolVk
+          bobsAddress = mkVkAddress networkId cardanoBobVk
+          -- carol's output
+          carolsOutput =
```
Should: remove pointless comment

ch1bo commented on hydra-cluster/src/Hydra/Cluster/Scenarios.hs:
```diff
@@ -339,6 +339,67 @@ canCloseWithLongContestationPeriod tracer workDir node@RunningNode{networkId} hy
+              TxOutDatumNone
+              ReferenceScriptNone
+      -- prepare fully balanced tx body
+      let eTxBody = buildRaw (fst <$> UTxO.pairs bobUTxO) [carolsOutput, bobsOutput] (Lovelace 165545)
```
This is very low-level as you have computed the fees and change yourself here. Instead, we should be using a bit more of the utility the `cardano-api` provides us. There is, for example the `buildTransaction` function, which uses the auto balancing which should make this a bit more robust:

```suggestion
      eTxBody <- buildTransaction networkId nodeSocket bobsAddress bobUTxO [] [carolsOutput]
```

ch1bo commented on hydra-cluster/src/Hydra/Cluster/Scenarios.hs:
```diff
@@ -339,6 +339,67 @@ canCloseWithLongContestationPeriod tracer workDir node@RunningNode{networkId} hy
+          carolsOutput =
+            TxOut
+              carolsAddress
+              (TxOutValue MultiAssetInBabbageEra $ lovelaceToValue $ Lovelace 2_000_000)
```
Should: use `mkTxOutValue`

ch1bo commented on hydra-cluster/src/Hydra/Cluster/Scenarios.hs:
```diff
@@ -339,6 +339,67 @@ canCloseWithLongContestationPeriod tracer workDir node@RunningNode{networkId} hy
+    refuelIfNeeded tracer node Alice 25_000_000
+    aliceChainConfig <- chainConfigFor Alice workDir nodeSocket [] $ UnsafeContestationPeriod 100
+    let hydraNodeId = 1
+    withHydraNode tracer aliceChainConfig workDir hydraNodeId aliceSk [] [1] hydraScriptsTxId $ \_ -> do
```
The withHydraNode interface is a bit contrived right now, we also need to provide all node ids and hence should re-use this variable:

```suggestion
    withHydraNode tracer aliceChainConfig workDir hydraNodeId aliceSk [] [hydraNodeId] hydraScriptsTxId $ \_ -> do
```

v0d1ch committed:
Change APIRestInputReceived to not include body and rename it to APIHTTPRequestReceived.

Also changed the definition to use dedicated types and simplify
construction in http server.

v0d1ch committed:
Remove APIHandshakeError

ch1bo reviewed (changes requested):
Not many changes needed anymore, although I'm puzzled by some of the renamings (and it makes reviewing this PR hard)

ch1bo committed:
Simplify HTTP handlers by not passing continuation

All branches of the handler were using respond, so we can factor this
out and make the scope of handler smaller.

ch1bo committed:
Remove imports with orphan instances from APIServerlog

This forced us to write instances for ByteString and Text, which is
actually fine as we could give a more precise definition what bytes/text
we expect here.

ch1bo committed:
Rename submitUserTx -> submitTx and simplify api.yaml

We can interpret the new endpoint as a basic submission endpoint and
indeed it is the same implementation as the Chain module uses itself
which gets exposed.

ch1bo committed:
Document possible response messages in api.ayml

AsyncAPI allows to specify multiple message kinds on a channel, which we
can use to enumerate the possible responses in HTTP endpoints.

ch1bo commented:
Only my comments in e2e tests remaining from my side

v0d1ch committed:
PR Review changes in Scenarios module

v0d1ch committed:
Remove leftover golden file and fix compilation
